crates/ruff/src/commands/format.rs

```diff
@@ -792,6 +1018,56 @@ impl Display for FormatCommandError {
     }
 }
 
+#[derive(Debug)]
+struct ModifiedRange {
+    unformatted: TextRange,
+    formatted: TextRange,
+}
+
+impl ModifiedRange {
+    /// Determine the range that differs between `unformatted` and `formatted`.
+    ///
+    /// If the two inputs are equal, the returned ranges will be empty.
+    fn new(unformatted: &SourceKind, formatted: &SourceKind) -> Self {
+        let unformatted = unformatted.source_code();
+        let formatted = formatted.source_code();
+
+        let start = unformatted
+            .char_indices()
+            .zip(formatted.chars())
+            .find_map(|((offset, old), new)| {
+                (old != new).then_some(TextSize::try_from(offset).unwrap())
+            })
+            // Fall back on the shorter text length if one of the strings is a strict prefix of the
+            // other (i.e. the zip iterator ended before finding a difference).
+            .unwrap_or_else(|| unformatted.text_len().min(formatted.text_len()));
```

Comment:
I wonder if a regular loop would have been easier here (similar to what you have below):

```rust
let mut prefix_length = TextSize::ZERO;

for (unformatted, formatted) in unformatted.chars().zip(formatted.chars()) {
    if unformatted != formatted {
        break;
    }

    prefix_length += unformatted.text_len();
}
```